File: 0337-house-robber-iii/0337-house-robber-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class pair{
        int isRob;
        int notRob;
        
        pair(int isRob,int notRob) {
            this.isRob = isRob;
            this.notRob = notRob;
        }
    }
    
     //Base Case
    //Recursive left and right call of faith
    //when my current node is roobed
    //when my current node is not robbed
    pair getMaxMoney(TreeNode node) {
        //Base Case
        if(node == null) return new pair(0,0);
        
        pair myAns = new pair(0,0);
        //Recursive left and right call of faith
        pair leftAns = getMaxMoney(node.left);
        pair rightAns = getMaxMoney(node.right);
        //when my current node is roobed
        myAns.isRob = leftAns.notRob + rightAns.notRob + node.val;
        //when my current node is not robbed
        myAns.notRob = Math.max(leftAns.notRob,leftAns.isRob) + 
                       Math.max(rightAns.notRob,rightAns.isRob);
        
        return myAns;
    }
    
    public int rob(TreeNode root) {
        pair rootAns = getMaxMoney(root);
        return Math.max(rootAns.isRob,rootAns.notRob);
    }
}

File: 0337-house-robber-iii/0337-house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    //<isRob,notRob>
    pair<int,int> getMaxMoney(TreeNode* node) {
        //Base Case
        if(!node) return {0,0};
        
        pair leftAns = getMaxMoney(node->left);
        pair rightAns = getMaxMoney(node->right);
        
        pair myAns = {0,0};
        myAns.first = node->val + leftAns.second + rightAns.second;
        myAns.second = max(leftAns.first,leftAns.second) + max(rightAns.first,rightAns.second);
        
        return myAns;
    }
    
    int rob(TreeNode* root) {
        pair rootAns = getMaxMoney(root);
        return max(rootAns.first,rootAns.second);
    }
};